stmt/stmts/insert.go

```diff
@@ -368,9 +358,36 @@ func getOnDuplicateUpdateColumns(assignList []*expression.Assignment, t table.Ta
 	return m, nil
 }
 
-func (s *InsertValues) initDefaultValues(ctx context.Context, t table.Table, row []interface{}, marked map[int]struct{}) (recordID int64, err error) {
+func (s *InsertValues) initDefaultValues(ctx context.Context, t table.Table, row []interface{}, marked map[int]struct{}) error {
 	var defaultValueCols []*column.Col
 	for i, c := range t.Cols() {
+		if mysql.HasPriKeyFlag(c.Flag) && t.Meta().PKIsHandle {
+			if row[i] != nil {
+				recordID, err := types.ToInt64(row[i])
+				if err != nil {
+					return errors.Trace(err)
+				}
+				row[i] = recordID
```

Comment:
It seems that above lines inside `if row[i] != nil` can be removed, as the type will be checked in `CastValues()`.